relayer/crates/starknet-integration-tests/src/tests/ics20.rs

```diff
@@ -589,16 +595,308 @@ fn test_starknet_ics20_contract() -> Result<(), Error> {
+                .deploy_contract(&erc20_class_hash, false, &calldata)
+                .await?
+        };
+
+        info!(
+            "deployed ERC20 contract to address: {:?}",
+            erc20_token_address
+        );
+
+        let balance_starknet_relayer_step_0 = starknet_chain
+            .query_token_balance(&erc20_token_address, address_starknet_relayer)
+            .await?;
+
+        info!("initial balance: {}", balance_starknet_relayer_step_0);
+
+        assert_eq!(
+            balance_starknet_relayer_step_0.quantity,
+            erc20_token_supply.into()
+        );
+
+        // {
+        //     let transfer_amount = transfer_quantity.into();
+
+        //     let message = starknet_chain.build_transfer_token_message(
+        //         address_starknet_b,
+        //         &StarknetAmount::new(transfer_amount, erc20_token_address),
+        //     )?;
+
+        //     let response = starknet_chain.send_message(message).await?;
+
+        //     info!("performed top-up of {transfer_quantity} tokens");
+
+        //     info!("response: {:?}", response);
+
+        //     let balance = starknet_chain
+        //         .query_token_balance(&erc20_token_address, address_starknet_b)
+        //         .await?;
+
+        //     info!("top-up balance: {}", balance);
+
+        //     assert_eq!(balance.quantity, transfer_quantity.into());
+        // }
+
+        {
+            // approve ics20 contract to spend the tokens for address_starknet_b
+            let call_data = cairo_encoding.encode(&product![
+                ics20_contract_address,
+                U256::from(transfer_quantity)
+            ])?;
+
+            let call = Call {
+                to: erc20_token_address,
+                selector: selector!("approve"),
+                calldata: call_data,
+            };
+
+            let message = StarknetMessage::new(call);
+
+            let response = starknet_chain.send_message(message).await?;
+
+            info!("ERC20 approve response: {:?}", response);
+        }
+
+        // submit ics20 transfer from Starknet to Cosmos
+
+        let starknet_ic20_packet_data = {
+            let denom = PrefixedDenom {
+                trace_path: vec![],
+                base: Denom::Native(erc20_token_address),
+            };
+
+            let amount = transfer_quantity.into();
+
+            let sender = Participant::Native(*address_starknet_relayer);
+
+            let receiver = Participant::External(address_cosmos_a.clone());
+
+            let memo = String::new();
+
+            TransferPacketData {
+                denom,
+                amount,
+                sender,
+                receiver,
+                memo,
+            }
+        };
+
+        // create ibc transfer message
+
+        let starknet_ics20_send_message = {
+            let current_starknet_time = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_secs();
```

Comment:
This isn’t the actual current Starknet time. Right? might be confusing. Couldn’t query the node for its timestamp?

Reply:
c8b370c